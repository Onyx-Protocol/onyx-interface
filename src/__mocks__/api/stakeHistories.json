{
  "data": [
    {
      "id": 291,
      "address": "0x0000000000000000000000000000000000000000",
      "tx_hash": "0xf97a7673403a643ad23edb9ffa4c60bf4192998aaf436782240b84be94d3aba1",
      "block_hash": "0x7d6991ef8f031e9cdc4ced2f924dee008e505c741f169101124aee16dad795f1",
      "type": 0,
      "amount": "11075000000000000000",
      "reward": "0",
      "price": "0.033695332045177356",
      "created_at": "2022-12-14T19:04:11.000Z",
      "updated_at": "2022-12-14T19:04:11.000Z"
    }
  ],
  "metadata": {
    "page": 1,
    "limit": 5,
    "totalItem": 1,
    "totalPage": 1
  }
}
