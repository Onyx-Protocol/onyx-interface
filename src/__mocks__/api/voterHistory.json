{
  "data": [
    {
      "voter": {
        "id": "40b3dcb3-f337-4713-ae4c-cae927351550",
        "address": "0xe864c9439eb042e969c59dd13d38cfdb28e16d73",
        "has_voted": 1,
        "support": 1,
        "votes": "232456299820100000000000000",
        "proposal_id": 6,
        "block_number": 14636926,
        "block_timestamp": 1650660975,
        "created_at": "2022-07-11T01:00:47.233Z",
        "updated_at": "2022-07-11T01:01:00.000Z"
      },
      "proposal": {
        "id": "1",
        "proposer": "0x1b7081943449db22e4d68c5bac45cfc85b44d3a7",
        "targets": [
          "0x08edf0f2af8672029eb445742b3b4072c6158df3"
        ],
        "values": [
          0
        ],
        "signatures": [
          "setDelay(uint256)"
        ],
        "callDatas": [
          "0x000000000000000000000000000000000000000000000000000000000000012c"
        ],
        "startBlock": 14636411,
        "endBlock": 14636441,
        "description": "\"# CIP 4 Pre Governance Activation\n Chain Improvement Proposal 4 (CIP-4) enables the Chain Governance module to update the Timelock smart contract for final mainnet testing purposes.\"",
        "state": "pending",
        "eta": 0,
        "forVotes": 109981950000000000000000000,
        "againstVotes": 0,
        "createdBlockNumber": 14636410,
        "createdBlockTimestamp": 1650654027,
        "createdTransactionHash": "0xb5c2ae6315b1904385063e2e2f90139a47eadb1622b3dcd4549abb054ea6c435",
        "queuedBlockNumber": null,
        "queuedBlockTimestamp": null,
        "queuedTransactionHash": null,
        "executedBlockNumber": null,
        "executedBlockTimestamp": null,
        "executedTransactionHash": null,
        "canceledBlockNumber": null,
        "canceledBlockTimestamp": null,
        "canceledTransactionHash": null
      }
    },
    {
      "voter": {
        "id": "7c29e70e-a985-4381-b193-c9394aed86dd",
        "address": "0xe864c9439eb042e969c59dd13d38cfdb28e16d73",
        "has_voted": 1,
        "support": 1,
        "votes": "232456299820100000000000000",
        "proposal_id": 5,
        "block_number": 14636725,
        "block_timestamp": 1650658028,
        "created_at": "2022-07-11T01:00:47.227Z",
        "updated_at": "2022-07-11T01:01:00.000Z"
      },
      "proposal": {
        "id": "5",
        "proposer": "0x1b7081943449db22e4d68c5bac45cfc85b44d3a7",
        "targets": [
          "0x08edf0f2af8672029eb445742b3b4072c6158df3"
        ],
        "values": [
          0
        ],
        "signatures": [
          "setDelay(uint256)"
        ],
        "callDatas": [
          "0x000000000000000000000000000000000000000000000000000000000000012c"
        ],
        "startBlock": 14636411,
        "endBlock": 14636441,
        "description": "\"# CIP 4 Pre Governance Activation\n Chain Improvement Proposal 4 (CIP-4) enables the Chain Governance module to update the Timelock smart contract for final mainnet testing purposes.\"",
        "state": "pending",
        "eta": 0,
        "forVotes": 109981950000000000000000000,
        "againstVotes": 0,
        "createdBlockNumber": 14636410,
        "createdBlockTimestamp": 1650654027,
        "createdTransactionHash": "0xb5c2ae6315b1904385063e2e2f90139a47eadb1622b3dcd4549abb054ea6c435",
        "queuedBlockNumber": null,
        "queuedBlockTimestamp": null,
        "queuedTransactionHash": null,
        "executedBlockNumber": null,
        "executedBlockTimestamp": null,
        "executedTransactionHash": null,
        "canceledBlockNumber": null,
        "canceledBlockTimestamp": null,
        "canceledTransactionHash": null
      }
    },
    {
      "voter": {
        "id": "1f55caa5-6dde-4eed-862c-3fedf31be63f",
        "address": "0xe864c9439eb042e969c59dd13d38cfdb28e16d73",
        "has_voted": 1,
        "support": 1,
        "votes": "232456299820100000000000000",
        "proposal_id": 2,
        "block_number": 14636476,
        "block_timestamp": 1650654753,
        "created_at": "2022-07-11T01:00:47.222Z",
        "updated_at": "2022-07-11T01:01:00.000Z"
      },
      "proposal": {
        "id": "2",
        "proposer": "0x1b7081943449db22e4d68c5bac45cfc85b44d3a7",
        "targets": [
          "0x08edf0f2af8672029eb445742b3b4072c6158df3"
        ],
        "values": [
          0
        ],
        "signatures": [
          "setDelay(uint256)"
        ],
        "callDatas": [
          "0x000000000000000000000000000000000000000000000000000000000000012c"
        ],
        "startBlock": 14636411,
        "endBlock": 14636441,
        "description": "\"# CIP 4 Pre Governance Activation\n Chain Improvement Proposal 4 (CIP-4) enables the Chain Governance module to update the Timelock smart contract for final mainnet testing purposes.\"",
        "state": "pending",
        "eta": 0,
        "forVotes": 109981950000000000000000000,
        "againstVotes": 0,
        "createdBlockNumber": 14636410,
        "createdBlockTimestamp": 1650654027,
        "createdTransactionHash": "0xb5c2ae6315b1904385063e2e2f90139a47eadb1622b3dcd4549abb054ea6c435",
        "queuedBlockNumber": null,
        "queuedBlockTimestamp": null,
        "queuedTransactionHash": null,
        "executedBlockNumber": null,
        "executedBlockTimestamp": null,
        "executedTransactionHash": null,
        "canceledBlockNumber": null,
        "canceledBlockTimestamp": null,
        "canceledTransactionHash": null
      }
    }
  ],
  "metadata": {
    "page": 1,
    "limit": 5,
    "totalItem": 3,
    "totalPage": 1
  }
}
