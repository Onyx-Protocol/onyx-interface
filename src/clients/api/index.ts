export { default as queryClient } from './queryClient';

// Mutations
export { default as enterMarkets } from './mutations/enterMarkets';
export * from './mutations/enterMarkets';
export { default as useEnterMarkets } from './mutations/enterMarkets/useEnterMarkets';

export { default as exitMarket } from './mutations/exitMarket';
export * from './mutations/exitMarket';
export { default as useExitMarket } from './mutations/exitMarket/useExitMarket';

export { default as approveToken } from './mutations/approveToken';
export * from './mutations/approveToken';
export { default as useApproveToken } from './mutations/approveToken/useApproveToken';

export { default as approveNftForAll } from './mutations/approveNftForAll';
export * from './mutations/approveNftForAll';
export { default as useApproveNftForAll } from './mutations/approveNftForAll/useApproveNftForAll';

export { default as supplyNonEth } from './mutations/supplyNonEth';
export * from './mutations/supplyNonEth';
export { default as useSupplyNonEth } from './mutations/supplyNonEth/useSupplyNonEth';
export * from './mutations/supplyNonEth/useSupplyNonEth';

export { default as supplyEth } from './mutations/supplyEth';
export * from './mutations/supplyEth';
export { default as useSupplyEth } from './mutations/supplyEth/useSupplyEth';
export * from './mutations/supplyEth/useSupplyEth';

export { default as supplyNFT } from './mutations/supplyNFT';
export * from './mutations/supplyNFT';
export { default as useSupplyNFT } from './mutations/supplyNFT/useSupplyNFT';
export * from './mutations/supplyNFT/useSupplyNFT';

export { default as redeem } from './mutations/redeem';
export * from './mutations/redeem';
export { default as useRedeem } from './mutations/redeem/useRedeem';

export { default as redeemNFT } from './mutations/redeemNFT';
export * from './mutations/redeemNFT';
export { default as useRedeemNFT } from './mutations/redeemNFT/useRedeemNFT';

export { default as repayNonEthOToken } from './mutations/repayNonEthOToken';
export * from './mutations/repayNonEthOToken';
export { default as useRepayNonEthOToken } from './mutations/repayNonEthOToken/useRepayNonEthOToken';

export { default as repayEth } from './mutations/repayEth';
export * from './mutations/repayEth';
export { default as useRepayEth } from './mutations/repayEth/useRepayEth';

export { default as redeemUnderlying } from './mutations/redeemUnderlying';
export * from './mutations/redeemUnderlying';
export { default as useRedeemUnderlying } from './mutations/redeemUnderlying/useRedeemUnderlying';

export { default as claimXcnReward } from './mutations/claimXcnReward';
export * from './mutations/claimXcnReward';
export { default as useClaimXcnReward } from './mutations/claimXcnReward/useClaimXcnReward';

export { default as borrowOToken } from './mutations/borrowOToken';
export * from './mutations/borrowOToken';
export { default as useBorrowOToken } from './mutations/borrowOToken/useBorrowOToken';

export { default as useRepayOToken } from './mutations/useRepayOToken';

export { default as createProposal } from './mutations/createProposal';
export * from './mutations/createProposal';
export { default as useCreateProposal } from './mutations/createProposal/useCreateProposal';

export { default as cancelProposal } from './mutations/cancelProposal';
export * from './mutations/cancelProposal';
export { default as useCancelProposal } from './mutations/cancelProposal/useCancelProposal';

export { default as executeProposal } from './mutations/executeProposal';
export * from './mutations/executeProposal';
export { default as useExecuteProposal } from './mutations/executeProposal/useExecuteProposal';

export { default as queueProposal } from './mutations/queueProposal';
export * from './mutations/queueProposal';
export { default as useQueueProposal } from './mutations/queueProposal/useQueueProposal';

export { default as castVote } from './mutations/vote/castVote';
export * from './mutations/vote/castVote';
export { default as useCastVote } from './mutations/vote/useCastVote';

export { default as registerProxy } from './mutations/registerProxy';
export * from './mutations/registerProxy';
export { default as useRegisterProxy } from './mutations/registerProxy/useRegisterProxy';

export { default as depositPunk } from './mutations/depositPunk';
export * from './mutations/depositPunk';
export { default as useDepositPunk } from './mutations/depositPunk/useDepositPunk';

export { default as mintWPunks } from './mutations/mintWPunks';
export * from './mutations/mintWPunks';
export { default as useMintWPunks } from './mutations/mintWPunks/useMintWPunks';

export { default as burnWPunks } from './mutations/burnWPunks';
export * from './mutations/burnWPunks';
export { default as useBurnWPunks } from './mutations/burnWPunks/useBurnWPunks';

export { default as swapTokens } from './mutations/swapTokens';
export * from './mutations/swapTokens';
export { default as useSwapTokens } from './mutations/swapTokens/useSwapTokens';

// Queries
export { default as getAssetsInAccount } from './queries/getAssetsInAccount';
export * from './queries/getAssetsInAccount';
export { default as useGetAssetsInAccount } from './queries/getAssetsInAccount/useGetAssetsInAccount';

export { default as getHypotheticalAccountLiquidity } from './queries/getHypotheticalAccountLiquidity';
export * from './queries/getHypotheticalAccountLiquidity';

export { default as getMarkets } from './queries/getMarkets';
export * from './queries/getMarkets';
export { default as useGetMarkets } from './queries/getMarkets/useGetMarkets';

export { default as getOTokenBalancesAll } from './queries/getOTokenBalancesAll';
export * from './queries/getOTokenBalancesAll';
export { default as useGetOTokenBalancesAll } from './queries/getOTokenBalancesAll/useGetOTokenBalancesAll';

export { default as getOTokenBalanceOf } from './queries/getOTokenBalanceOf';
export * from './queries/getOTokenBalanceOf';
export { default as useGetOTokenBalanceOf } from './queries/getOTokenBalanceOf/useGetOTokenBalanceOf';

export { default as getXcnReward } from './queries/getXcnReward';
export * from './queries/getXcnReward';
export { default as useGetXcnReward } from './queries/getXcnReward/useGetXcnReward';

export { default as getAllowance } from './queries/getAllowance';
export * from './queries/getAllowance';
export { default as useGetAllowance } from './queries/getAllowance/useGetAllowance';

export { default as getIsApprovedForAll } from './queries/getIsApprovedForAll';
export * from './queries/getIsApprovedForAll';
export { default as useGetIsApprovedForAll } from './queries/getIsApprovedForAll/useGetIsApprovedForAll';

export { default as getBalanceOf } from './queries/getBalanceOf';
export * from './queries/getBalanceOf';
export { default as useGetBalanceOf } from './queries/getBalanceOf/useGetBalanceOf';

export { default as getTokenBalances } from './queries/getTokenBalances';
export * from './queries/getTokenBalances';
export { default as useGetTokenBalances } from './queries/getTokenBalances/useGetTokenBalances';

export { default as useGetUserMarketInfo } from './queries/useGetUserMarketInfo';

export { default as useGetTreasuryTotals } from './queries/useGetTreasuryTotals';

export { default as getMarketHistory } from './queries/getMarketHistory';
export * from './queries/getMarketHistory';
export { default as useGetMarketHistory } from './queries/getMarketHistory/useGetMarketHistory';

export { default as getOTokenCash } from './queries/getOTokenCash';
export * from './queries/getOTokenCash';
export { default as useGetOTokenCash } from './queries/getOTokenCash/useGetOTokenCash';

export { default as getOTokenInterestRateModel } from './queries/getOTokenInterestRateModel';
export * from './queries/getOTokenInterestRateModel';
export { default as useGetOTokenInterestRateModel } from './queries/getOTokenInterestRateModel/useGetOTokenInterestRateModel';

export { default as getOTokenApySimulations } from './queries/getOTokenApySimulations';
export * from './queries/getOTokenApySimulations';
export { default as useGetOTokenApySimulations } from './queries/getOTokenApySimulations/useGetOTokenApySimulations';

export { default as getCurrentVotes } from './queries/getCurrentVotes';
export * from './queries/getCurrentVotes';
export { default as useGetCurrentVotes } from './queries/getCurrentVotes/useGetCurrentVotes';

export { default as getStakingInfos } from './queries/getStakingInfos';
export * from './queries/getStakingInfos';
export { default as useGetStakingInfos } from './queries/getStakingInfos/useGetStakingInfos';

export { default as getStakingApy } from './queries/getStakingApy';
export * from './queries/getStakingApy';
export { default as useGetStakingApy } from './queries/getStakingApy/useGetStakingApy';

export { default as getTransactions } from './queries/getTransactions';
export * from './queries/getTransactions';
export { default as useGetTransactions } from './queries/getTransactions/useGetTransactions';

export { default as getDailyXcn } from './queries/getDailyXcn';
export * from './queries/getDailyXcn';
export { default as useGetDailyXcn } from './queries/getDailyXcn/useGetDailyXcn';

export { default as getProposals } from './queries/getProposals';
export * from './queries/getProposals';
export { default as useGetProposals } from './queries/getProposals/useGetProposals';

export * from './queries/getStakeHistories';
export { default as useGetStakeHistories } from './queries/getStakeHistories/useGetStakeHistories';

export { default as getProposal } from './queries/getProposals/getProposal';
export * from './queries/getProposals/getProposal';
export { default as useGetProposal } from './queries/getProposals/useGetProposal';

export { default as getVoteReceipt } from './queries/getVoteReceipt';
export * from './queries/getVoteReceipt';
export { default as useGetVoteReceipt } from './queries/getVoteReceipt/useGetVoteReceipt';

export { default as getVoters } from './queries/getVoters';
export * from './queries/getVoters';
export { default as useGetVoters } from './queries/getVoters/useGetVoters';

export { default as getVoterDetails } from './queries/getVoterDetails';
export * from './queries/getVoterDetails';
export { default as useGetVoterDetails } from './queries/getVoterDetails/useGetVoterDetails';

export { default as getVoterHistory } from './queries/getVoterHistory';
export * from './queries/getVoterHistory';
export { default as useGetVoterHistory } from './queries/getVoterHistory/useGetVoterHistory';

export { default as getVoterAccounts } from './queries/getVoterAccounts';
export * from './queries/getVoterAccounts';
export { default as useGetVoterAccounts } from './queries/getVoterAccounts/useGetVoterAccounts';

export { default as getProposalThreshold } from './queries/getProposalThreshold';
export * from './queries/getProposalThreshold';
export { default as useGetProposalThreshold } from './queries/getProposalThreshold/useGetProposalThreshold';

export { default as getProposalState } from './queries/getProposalState';
export * from './queries/getProposalState';
export { default as useGetProposalState } from './queries/getProposalState/useGetProposalState';

export { default as getLatestProposalIdByProposer } from './queries/getLatestProposalIdByProposer';
export * from './queries/getLatestProposalIdByProposer';
export { default as useGetLatestProposalIdByProposer } from './queries/getLatestProposalIdByProposer/useGetLatestProposalIdByProposer';

export { default as getBlockNumber } from './queries/getBlockNumber';
export * from './queries/getBlockNumber';
export { default as useGetBlockNumber } from './queries/getBlockNumber/useGetBlockNumber';

export { default as getProposalEta } from './queries/getProposalEta';
export * from './queries/getProposalEta';
export { default as useGetProposalEta } from './queries/getProposalEta/useGetProposalEta';

export { default as getProxies } from './queries/getProxies';
export * from './queries/getProxies';
export { default as useGetProxies } from './queries/getProxies/useGetProxies';

export { default as getOwnedPunkIds } from './queries/getOwnedPunkIds';
export * from './queries/getOwnedPunkIds';
export { default as useGetOwnedPunkIds } from './queries/getOwnedPunkIds/useGetOwnedPunkIds';

export { default as getOwnedWPunksIds } from './queries/getOwnedWPunksIds';
export * from './queries/getOwnedWPunksIds';
export { default as useGetOwnedWPunksIds } from './queries/getOwnedWPunksIds/useGetOwnedWPunksIds';

export { default as liquidateWithSingleRepay } from './mutations/liquidateWithSingleRepay';
export * from './mutations/liquidateWithSingleRepay';
export { default as useLiquidateWithSingleRepay } from './mutations/liquidateWithSingleRepay/useLiquidateWithSingleRepay';

export { default as liquidateBorrow } from './mutations/liquidateBorrow';
export * from './mutations/liquidateBorrow';
export { default as useLiquidateBorrow } from './mutations/liquidateBorrow/useLiquidateBorrow';

export { default as stakeXcn } from './mutations/stakeXcn';
export * from './mutations/stakeXcn';
export { default as useStakeXcn } from './mutations/stakeXcn/useStakeXcn';

export { default as withdrawXcn } from './mutations/withdrawXcn';
export * from './mutations/withdrawXcn';
export { default as useWithdrawXcn } from './mutations/withdrawXcn/useWithdrawXcn';

export { default as claimXcn } from './mutations/claimXcn';
export * from './mutations/claimXcn';
export { default as useClaimXcn } from './mutations/claimXcn/useClaimXcn';

export { default as getUniSwapPairs } from './queries/getUniSwapPairs';
export * from './queries/getUniSwapPairs';
export { default as useGetUniSwapPairs } from './queries/getUniSwapPairs/useGetUniSwapPairs';
