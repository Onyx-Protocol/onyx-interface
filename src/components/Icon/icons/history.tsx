import * as React from 'react';
import { SVGProps } from 'react';

const SvgHistory = (props: SVGProps<SVGSVGElement>) => (
  <svg
    width="12"
    height="15"
    viewBox="0 0 12 15"
    fill="none"
    xmlns="http://www.w3.org/2000/svg"
    {...props}
  >
    <path
      d="M9.5 15H2.5C1.90367 14.9987 1.33214 14.7612 0.910469 14.3395C0.488798 13.9179 0.251321 13.3463 0.25 12.75V5.75C0.251321 5.15367 0.488798 4.58214 0.910469 4.16047C1.33214 3.7388 1.90367 3.50132 2.5 3.5H9.5C10.0963 3.50132 10.6679 3.7388 11.0895 4.16047C11.5112 4.58214 11.7487 5.15367 11.75 5.75V12.75C11.7487 13.3463 11.5112 13.9179 11.0895 14.3395C10.6679 14.7612 10.0963 14.9987 9.5 15ZM2.5 4.5C2.16848 4.5 1.85054 4.6317 1.61612 4.86612C1.3817 5.10054 1.25 5.41848 1.25 5.75V12.75C1.25 13.0815 1.3817 13.3995 1.61612 13.6339C1.85054 13.8683 2.16848 14 2.5 14H9.5C9.83152 14 10.1495 13.8683 10.3839 13.6339C10.6183 13.3995 10.75 13.0815 10.75 12.75V5.75C10.75 5.41848 10.6183 5.10054 10.3839 4.86612C10.1495 4.6317 9.83152 4.5 9.5 4.5H2.5Z"
      fill="currentColor"
    />
    <path
      d="M9.5 2.75H2.5C2.36739 2.75 2.24021 2.69732 2.14645 2.60355C2.05268 2.50979 2 2.38261 2 2.25C2 2.11739 2.05268 1.99021 2.14645 1.89645C2.24021 1.80268 2.36739 1.75 2.5 1.75H9.5C9.63261 1.75 9.75979 1.80268 9.85355 1.89645C9.94732 1.99021 10 2.11739 10 2.25C10 2.38261 9.94732 2.50979 9.85355 2.60355C9.75979 2.69732 9.63261 2.75 9.5 2.75Z"
      fill="currentColor"
    />
    <path
      d="M8.33504 1H3.66504C3.53243 1 3.40525 0.947322 3.31149 0.853553C3.21772 0.759785 3.16504 0.632608 3.16504 0.5C3.16504 0.367392 3.21772 0.240215 3.31149 0.146447C3.40525 0.0526784 3.53243 0 3.66504 0H8.33504C8.46765 0 8.59482 0.0526784 8.68859 0.146447C8.78236 0.240215 8.83504 0.367392 8.83504 0.5C8.83504 0.632608 8.78236 0.759785 8.68859 0.853553C8.59482 0.947322 8.46765 1 8.33504 1Z"
      fill="currentColor"
    />
    <path
      d="M8.33504 6.83496H3.66504C3.53243 6.83496 3.40525 6.78228 3.31149 6.68851C3.21772 6.59475 3.16504 6.46757 3.16504 6.33496C3.16504 6.20235 3.21772 6.07518 3.31149 5.98141C3.40525 5.88764 3.53243 5.83496 3.66504 5.83496H8.33504C8.46765 5.83496 8.59482 5.88764 8.68859 5.98141C8.78236 6.07518 8.83504 6.20235 8.83504 6.33496C8.83504 6.46757 8.78236 6.59475 8.68859 6.68851C8.59482 6.78228 8.46765 6.83496 8.33504 6.83496Z"
      fill="currentColor"
    />
  </svg>
);

export default SvgHistory;
