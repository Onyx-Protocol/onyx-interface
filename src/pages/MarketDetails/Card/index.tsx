/* eslint-disable no-nested-ternary */

/** @jsxImportSource @emotion/react */
import Paper from '@mui/material/Paper';
import Typography from '@mui/material/Typography';
import React from 'react';

import { Stat } from '../types';
import { useStyles } from './styles';

export interface Legend {
  label: string;
  color: string;
}

export interface CardProps {
  title: string;
  legends?: Legend[];
  stats?: Stat[];
  className?: string;
  testId?: string;
}

const Card: React.FC<CardProps> = ({
  children,
  title,
  legends = [],
  stats = [],
  className,
  testId,
}) => {
  const styles = useStyles();

  return (
    <Paper className={className} data-testid={testId}>
      <div css={styles.header}>
        <h4 css={styles.title}>{title}</h4>

        {legends.length > 0 && (
          <div css={styles.row}>
            {legends.map(legend => (
              <div css={styles.legend} key={`card-${title}-legend-${legend.label}`}>
                <div css={styles.getLegendColorIndicator({ color: legend.color })} />

                <Typography css={styles.legendLabel} variant="small2">
                  {legend.label}
                </Typography>
              </div>
            ))}
          </div>
        )}
      </div>

      {stats.length > 0 && (
        <div css={styles.row}>
          {stats.map((stat, index) => (
            <div css={styles.stat} key={`card-${title}-legend-${stat.label}`}>
              <Typography variant="small2" component="div" css={styles.statLabel}>
                {stat.label}
              </Typography>

              <Typography
                variant="h4"
                css={styles.statValue}
                style={{
                  color:
                    index === 0
                      ? 'white'
                      : title === 'Supply info'
                      ? '#18DF8B'
                      : index === 1
                      ? '#E93D44'
                      : index === 2
                      ? '#18DF8B'
                      : 'white',
                }}
              >
                {stat.value}
              </Typography>
            </div>
          ))}
        </div>
      )}

      {children}
    </Paper>
  );
};

export default Card;
