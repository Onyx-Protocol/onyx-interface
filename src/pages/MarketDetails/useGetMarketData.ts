import BigNumber from 'bignumber.js';
import React from 'react';
import { Token } from 'types';
import {
  convertPercentageFromSmartContract,
  convertWeiToTokens,
  unsafelyGetToken,
  unsafelyGetOToken,
} from 'utilities';

import { useGetMarkets, useGetOTokenCash } from 'clients/api';
import { BLOCKS_PER_DAY } from 'constants/ethereum';
import { COMPOUND_MANTISSA } from 'constants/compoundMantissa';
import { TOKENS } from 'constants/tokens';

const useGetMarketData = ({ oTokenId }: { oTokenId: Token['id'] }) => {
  const { data: oTokenCashData } = useGetOTokenCash({
    oTokenId,
  });

  const { data: getMarketData } = useGetMarkets();
  const assetMarket = (getMarketData?.markets || []).find(market => market.id === oTokenId);

  return React.useMemo(() => {
    const token = unsafelyGetToken(oTokenId);
    const oToken = unsafelyGetOToken(oTokenId);
    const totalBorrowBalanceCents = assetMarket && +assetMarket.totalBorrowsUsd * 100;
    const totalSupplyBalanceCents = assetMarket && +assetMarket.totalSupplyUsd * 100;
    const borrowApyPercentage = assetMarket?.borrowApy;
    const supplyApyPercentage = assetMarket?.supplyApy;
    const borrowDistributionApyPercentage = assetMarket && +assetMarket.borrowXcnApy;
    const supplyDistributionApyPercentage = assetMarket && +assetMarket.supplyXcnApy;
    const tokenPriceDollars = assetMarket?.tokenPrice;
    const liquidityCents = assetMarket && new BigNumber(assetMarket.liquidity).multipliedBy(100);
    const supplierCount = assetMarket?.supplierCount;
    const borrowerCount = assetMarket?.borrowerCount;
    const borrowCapTokens = assetMarket && new BigNumber(assetMarket.borrowCap);
    const supplyCapTokens = assetMarket && new BigNumber(assetMarket.supplyCap);
    const mintedTokens = assetMarket && new BigNumber(assetMarket.totalSupply2);
    const reserveFactorMantissa = assetMarket && new BigNumber(assetMarket.reserveFactor);

    const dailyDistributionXcn =
      assetMarket &&
      convertWeiToTokens({
        valueWei: new BigNumber(assetMarket.supplierDailyXcn).plus(
          assetMarket.borrowerDailyXcn,
        ),
        token: TOKENS.xcn,
      });

    const formattedSupplyRatePerBlock =
      assetMarket &&
      new BigNumber(assetMarket.supplyRatePerBlock).dividedBy(COMPOUND_MANTISSA).toNumber();

    const formattedBorrowRatePerBlock =
      assetMarket &&
      new BigNumber(assetMarket.borrowRatePerBlock).dividedBy(COMPOUND_MANTISSA).toNumber();

    // Calculate daily interests for suppliers and borrowers. Note that we don't
    // use BigNumber to calculate these values, as this would slow down
    // calculation a lot while the end result doesn't need to be extremely
    // precise
    const dailySupplyingInterestsCents =
      assetMarket &&
      formattedSupplyRatePerBlock &&
      // prettier-ignore
      +assetMarket.totalSupplyUsd * (((1 + formattedSupplyRatePerBlock) ** BLOCKS_PER_DAY) - 1) *
      // Convert to cents
      100;

    const dailyBorrowingInterestsCents =
      assetMarket &&
      formattedBorrowRatePerBlock &&
      // prettier-ignore
      +assetMarket.totalBorrowsUsd * (((1 + formattedBorrowRatePerBlock) ** BLOCKS_PER_DAY) - 1)
        // Convert to cents
        * 100;

    const reserveFactor =
      assetMarket && convertPercentageFromSmartContract(assetMarket.reserveFactor);

    const collateralFactor =
      assetMarket && convertPercentageFromSmartContract(assetMarket.collateralFactor);

    const reserveTokens =
      assetMarket &&
      convertWeiToTokens({
        valueWei: new BigNumber(assetMarket.totalReserves),
        token,
      });

    const exchangeRateOTokens =
      assetMarket &&
      new BigNumber(1).div(
        new BigNumber(assetMarket.exchangeRate).div(
          new BigNumber(10).pow(18 + token.decimals - oToken.decimals),
        ),
      );

    let currentUtilizationRate: number | undefined;
    if (oTokenCashData?.cashWei && assetMarket && reserveTokens) {
      const oTokenCashTokens = convertWeiToTokens({
        valueWei: oTokenCashData.cashWei,
        token,
      });

      currentUtilizationRate = new BigNumber(assetMarket.totalBorrows2)
        .div(oTokenCashTokens.plus(assetMarket.totalBorrows2).minus(reserveTokens))
        .multipliedBy(100)
        .dp(0)
        .toNumber();
    }

    return {
      totalBorrowBalanceCents,
      totalSupplyBalanceCents,
      borrowApyPercentage,
      supplyApyPercentage,
      borrowDistributionApyPercentage,
      supplyDistributionApyPercentage,
      tokenPriceDollars,
      liquidityCents,
      supplierCount,
      borrowerCount,
      borrowCapTokens,
      supplyCapTokens,
      mintedTokens,
      dailyDistributionXcn,
      dailySupplyingInterestsCents,
      dailyBorrowingInterestsCents,
      reserveFactor,
      collateralFactor,
      reserveTokens,
      exchangeRateOTokens,
      currentUtilizationRate,
      reserveFactorMantissa,
    };
  }, [JSON.stringify(assetMarket), oTokenCashData?.cashWei.toFixed()]);
};

export default useGetMarketData;
