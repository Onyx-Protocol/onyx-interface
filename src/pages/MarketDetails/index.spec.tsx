import { waitFor } from '@testing-library/react';
import BigNumber from 'bignumber.js';
import { createMemoryHistory } from 'history';
import React from 'react';

import { marketSnapshots } from '__mocks__/models/marketSnapshots';
import { markets } from '__mocks__/models/markets';
import { oTokenApySimulations } from '__mocks__/models/oTokenApySimulations';
import { getMarketHistory, getMarkets, getOTokenApySimulations } from 'clients/api';
import renderComponent from 'testUtils/renderComponent';

import MarketDetails from '.';
import TEST_IDS from './testIds';

const fakeOTokenId = 'wbtc';

jest.mock('clients/api');

describe('pages/MarketDetails', () => {
  beforeEach(() => {
    (getMarketHistory as jest.Mock).mockImplementation(() => ({
      marketSnapshots,
    }));
    (getMarkets as jest.Mock).mockImplementation(() => ({
      markets,
      dailyXcnWei: new BigNumber(0),
    }));
    (getOTokenApySimulations as jest.Mock).mockImplementation(() => ({
      apySimulations: oTokenApySimulations,
    }));
  });

  it('renders without crashing', () => {
    const fakeHistory = createMemoryHistory();
    renderComponent(
      <MarketDetails
        history={fakeHistory}
        location="/"
        match={{
          params: {
            oTokenId: fakeOTokenId,
          },
          isExact: true,
          path: '/:oTokenId',
          url: '',
        }}
      />,
    );
  });

  it('fetches market details and displays them correctly', async () => {
    const fakeHistory = createMemoryHistory();
    const { getByTestId } = renderComponent(
      <MarketDetails
        history={fakeHistory}
        location="/"
        match={{
          params: {
            oTokenId: fakeOTokenId,
          },
          isExact: true,
          path: '/:oTokenId',
          url: '',
        }}
      />,
    );

    // Check supply info displays correctly
    await waitFor(() => expect(getByTestId(TEST_IDS.supplyInfo).textContent).toMatchSnapshot());
    // Check borrow info displays correctly
    expect(getByTestId(TEST_IDS.borrowInfo).textContent).toMatchSnapshot();
    // Check interest rate model displays correctly
    expect(getByTestId(TEST_IDS.interestRateModel).textContent).toMatchSnapshot();
    // Check market info displays correctly
    expect(getByTestId(TEST_IDS.marketInfo).textContent).toMatchSnapshot();
  });
});
