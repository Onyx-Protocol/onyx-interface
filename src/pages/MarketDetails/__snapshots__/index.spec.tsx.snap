// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/MarketDetails fetches market details and displays them correctly 1`] = `"Supply infoSupply APYTotal supply$34.35MAPY1.81%Distribution APY1.61%"`;

exports[`pages/MarketDetails fetches market details and displays them correctly 2`] = `"Borrow infoBorrow APYTotal borrow$12.32MAPY0%Distribution APY4.54%"`;

exports[`pages/MarketDetails fetches market details and displays them correctly 3`] = `"Interest Rate ModelUtilization rateBorrow APYSupply APY"`;

exports[`pages/MarketDetails fetches market details and displays them correctly 4`] = `"Market infoPrice$30,046.80Market liquidity$22,080,059.44# of suppliers415# of borrowers51Supply capUncappedBorrow capUncappedDaily supplying interests$424.31Daily borrowing interests$471.47Daily XCN distributed3,000Reserves17,660,902,508.1563 WBTCReserve factor10%Collateral factor80%oBTC minted53,877.5661Exchange rate1 WBTC=0 oBTC"`;
